File: Programmers/Practice/Level 1/failure.cpp
//2020.06.06
//프로그래머스 실패율 (2019 KAKAO BLIND RECRUITMENT)
//By Mincastle

#include <vector>
#include <algorithm>
#include <functional>

using namespace std;

vector<int> solution(int N, vector<int> stages) {
    vector<int> answer;
    vector<double> failure;

    sort(stages.begin(), stages.end(), greater<int>());

    for (int i = 0; i <= N; i++) {
        int cnt = 0;
        double total = stages.size();

        if (stages.empty()) {
            failure.push_back(0);
            continue;
        }

        while (stages.back() == i + 1) {
            stages.pop_back();
            cnt++;
        }
        failure.push_back((total - stages.size()) / total);
    }

    for (int i = 0; i < failure.size(); i++) {
        int idx = max_element(failure.begin(), failure.end()) - failure.begin() + 1;

        if (idx <= N) answer.push_back(idx);
        failure[idx - 1] = -1;
    }

    return answer;
}

int main(void) {
    vector<int> stages = {2,1,2,6,2,4,3,3};

    solution(5, stages);

    return 0;
}

File: Kakao/failure.py
# 실패율
def solution(N, stages):
    user = [0 for i in range(N + 1)]
    failure = {}

    for s in stages:
        user[s - 1] += 1

    for i in range(N - 1, -1, -1):
        failure[i + 1] = user[i] / (user[i] + user[i + 1]) if user[i] + user[i + 1] != 0 else 0
        user[i] += user[i + 1]

    answer = sorted(failure.items(), key=(lambda x: x[0]))
    answer = sorted(answer, key=(lambda x: x[1]), reverse=True)
    answer = list(map(lambda x: x[0], answer))

    return answer


print(solution(5, [2, 1, 2, 6, 2, 4, 3, 3]))
